File: Programmers/Lv. 3/베스트앨범.py
# 장르별 가장 많이 재생된 노래 2개씩 모아 베스트 앨범 출시
# 노래는 고유 번호로 구분
# 속한 노래가 많이 재생된 장르 먼저 수록
# 장르 내 많이 재생된 노래 먼저 수록, 만약 재생 횟수 같으면 고유 번호 낮은 노래 먼저 수록
    
def solution(genres, plays):
    answer = []
    n = len(genres)
    
    songs = {}
    playCnt = {}
    
    for i in range(n):
        if genres[i] not in songs.keys():
            songs[genres[i]] = [(i, plays[i])]
            playCnt[genres[i]] = plays[i]
        else:
            songs[genres[i]].append((i, plays[i]))
            playCnt[genres[i]] += plays[i]
            
    for genre in sorted(playCnt, key=lambda x : playCnt[x], reverse=True):
        songs[genre].sort(key=lambda x : (-x[1], x[0]))
        try:
            answer.append(songs[genre][0][0])
            answer.append(songs[genre][1][0])
        except:
            pass
    
    return answer

File: Programmers/Lv. 3/베스트앨범.cpp
/*
장르를 나타내는 문자열 배열 genres
노래별 재생 횟수를 나타내는 정수 배열 plays
베스트 앨범에 들어갈 노래의 고유 번호를 순서대로 return
*/

#include <string>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

bool compareByValue(pair<string, int>& a, pair<string, int>& b) {
    return a.second > b.second;
}

bool compareByPlay(pair<int, int>& a, pair<int, int>& b) {
    if (a.second == b.second)
    {
        return a.first < b.first;
    }
    return a.second > b.second;
}

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    
    map<string, int> cnt_by_genres;
    map<string, vector<pair<int, int>>> idx_type;
    
    for (int i = 0; i < genres.size(); i++)
    {
        if (cnt_by_genres.count(genres[i]) == 0)
        {
            cnt_by_genres[genres[i]] = 0;
            idx_type[genres[i]] = {};
        }
        cnt_by_genres[genres[i]] += plays[i];
        idx_type[genres[i]].push_back({i, plays[i]});
    }
    
    vector<pair<string,int>> v_cnt_by_genres(cnt_by_genres.begin(), cnt_by_genres.end());
    
    sort(v_cnt_by_genres.begin(), v_cnt_by_genres.end(), compareByValue);
    for (auto pair : v_cnt_by_genres)
    {
        auto tmp = idx_type[pair.first];
        sort(tmp.begin(), tmp.end(), compareByPlay);
        answer.push_back(tmp[0].first);
        if (tmp.size() > 1)
        {
            answer.push_back(tmp[1].first);
        }
    } 
    
    return answer;
}